### PR TITLE
[RFR]: Support server validation

### PR DESCRIPTION
**Pull request by tkvw**
This PR introduces a validation provider to transform validation errors returned from the server to `redux-form` [SubmissionErrors](https://redux-form.com/7.2.0/docs/api/submissionerror.md/). This allows to inform the user of validation errors returned from the server.

The signature is simple: 
```
validationProvider = error => {Promise}
```
It should reject when it cannot convert the error (the default implementation rejects all errors) and resolve with the errors when it succeeds. 

Example: 
```
httpValidationProvider = error => 422 === error.status? Promise.resolve(error): Promise.reject(error);
```

An interesting sideeffect of this PR is, that the `redux-form` actions are triggered correctly, because a Promise is returned from `handleSubmit` @see: https://redux-form.com/7.2.0/docs/api/reduxform.md/#-code-onsubmit-function-code-optional-. 

All errors are translated before displaying. This allows to work with message labels returned from the server. 

For clarity the `redux-form` errors object looks like: 
```
{
    _error: <genericerror>,
    field_a:<error_label>,
    field_b:<error_label>,
}
```

### CODE REVIEW

**Comment by tkvw** [comment]
This PR was getting unreadable, so I put some effort into refactoring my commits. Now each commit will tell the story of the entire PR, so it can be easier understood.

File: packages/react-admin/src/mui/form/GenericFormError.js [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+import { withStyles } from 'material-ui/styles';
+
+const styles = theme => ({
+    root: {
+        backgroundColor: theme.palette.background.default,
+        padding: 2 * theme.spacing.unit,
+    },
+});
+
+const GenericFormError = ({ translate, error, classes, ...rest }) =>
```
Comment (djhi) [inline packages/react-admin/src/mui/form/GenericFormError.js]: Can you add a `className` prop as well and merge it with `classes.root` using `classnames` as we do in other components ?

**Comment by fzaninotto** [comment]
I have an alternative to propose: what if developers could throw `SubmissionError` directly in their `dataProvider`? It's just a matter of filtering errors by type in the `crudFetch` sage, and letting redux-form errors bubble up. That way, errors from the data provider would be displayed in the form. You can keep the translation part of the PR.

I suggest that because I feel uncomfortable adding a new level of complexity : you add a new validation phase before submission, but with most data providers, it's done by the same controller. I prefer to leave the implementation details in the hands of the user.

What do you think?

**Comment by tkvw** [comment]
Changed the implementation so the `dataProvider` can throw a `SubmissionError`. Added changes requested by @djhi and superseed https://github.com/marmelab/admin-on-rest/pull/1349. Added a `requestIdentifier` to `resolve/reject` the correct form promise.

File: packages/react-admin/src/mui/form/SanitizedForm.js [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+    autofill,
+    submit,
+    resource,
+    redirect,
+    array,
+    ...rest
+}) => rest;
+
+const SanitizedForm = ({ renderLayout, ...props }) =>
+    renderLayout ? (
```
Comment (djhi) [inline packages/react-admin/src/mui/form/SanitizedForm.js]: Can you rename it `render` ?

File: packages/react-admin/src/mui/form/SanitizedForm.js [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    translate,
+    autofill,
+    submit,
+    resource,
+    redirect,
+    array,
+    ...rest
+}) => rest;
+
+const SanitizedForm = ({ renderLayout, ...props }) =>
```
Comment (djhi) [inline packages/react-admin/src/mui/form/SanitizedForm.js]: Can you rename `renderLayout` to `render` ?

File: packages/react-admin/src/mui/form/SanitizedForm.js [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+    renderLayout ? (
+        renderLayout({
+            ...props,
+            getFormProps: sanitizeFormProps,
```
Comment (djhi) [inline packages/react-admin/src/mui/form/SanitizedForm.js]: The name implies it is a function. Can you either make it a function or rename it `formProps` ? I think the second is better as a function does not bring anything here

File: packages/react-admin/src/mui/form/SanitizedForm.js [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+        })
+    ) : (
+        <form {...sanitizeFormProps(props)} />
```
Comment (djhi) [inline packages/react-admin/src/mui/form/SanitizedForm.js]: Wouldn't it be better to only have the `render` prop and provide a default for it returning this ?

**Comment by tkvw** [comment]
Moved form modifications to https://github.com/marmelab/admin-on-rest/pull/1389

**Comment by fzaninotto** [comment]
This PR was closed by GitHub when I renamed the repository from "admin-on-rest" to "react-admin", and the `next` branch `master`.

Would you mind reopening a new one with the same changes?

**Pull request closed without merging**